ciao-cli/template.go

```diff
@@ -213,3 +215,120 @@ func createTemplate(name, tmplSrc string) *template.Template {
 
 	return t
 }
+
+func exportedFields(typ reflect.Type) bool {
+	var i int
+	for i = 0; i < typ.NumField(); i++ {
+		if typ.Field(i).PkgPath == "" {
+			break
+		}
+	}
+
+	return i < typ.NumField()
+}
+
+func ignoreKind(kind reflect.Kind) bool {
+	return (kind == reflect.Chan) || (kind == reflect.Invalid)
+}
+
+func generateStruct(buf io.Writer, dummy reflect.Type) {
+	for i := 0; i < dummy.NumField(); i++ {
+		field := dummy.Field(i)
+		if field.PkgPath != "" || ignoreKind(field.Type.Kind()) {
+			continue
+		}
+		fmt.Fprintf(buf, "%s ", field.Name)
+		tag := ""
+		if field.Tag != "" {
+			tag = fmt.Sprintf("`%s`", field.Tag)
+		}
+		generateUsage(buf, field.Type, tag)
+	}
+}
+
+// This function is not currently used.  The idea was to annotate
+// types with their user defined typenames, if such names existed.
+// This would allow users to know whether they could invoke
+// methods on those types.  Unfortunately, I couldn't figure out
+// a nice way of displaying this information, so it's disabled for
+// now.
+func getFullTypeName(typ reflect.Type) string {
+	if typ.Name() == "" {
+		return ""
+	}
+	pkg := typ.PkgPath()
+	if pkg != "" {
+		li := strings.LastIndex(pkg, "/")
+		if li != -1 {
+			pkg = pkg[li+1:]
+		}
+		pkg = fmt.Sprintf("%s.", pkg)
+	}
+	return fmt.Sprintf("%s%s", pkg, typ.Name())
+}
+
+func generateUsage(buf io.Writer, dummy reflect.Type, tag string) {
```

Comment:
Code has been changed.  I use i for interface{} and typ for reflect.Type to match the reflect package.